samples/coop_synchronization/src/main.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2012-2014 Wind River Systems, Inc.
+ * Copyright (c) 2023 Advanced Micro Devices, Inc. (AMD)
+ * Copyright (c) 2023 Alp Sayin <alpsayin@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
+
+/*
+ * The coop synchronization demo has two threads that utilize semaphores and k_yield
+ * to take turns printing a greeting message with a somewhat uncontrolled rate.
+ * The rate is controlled via a busy waiting loop. The advantage of this demo is that
+ * it'll work even if there are no timing utilities provided by the architecture.
+ */
+
+/* size of stack area used by each thread */
+#define STACKSIZE 1024
+
+/* number of NOPs to churn through for busy waiting */
+#if defined(CONFIG_TIMING_FUNCTIONS)
+#define NUM_NOPS() timing_freq_get()
+#else
+#define NUM_NOPS() (1000000)
+#endif
+
+/* Some locks to take and release to make things interesting */
+static struct k_spinlock lock1, lock2;
+
+/*
+ * @param my_name      thread identification string
+ * @param my_sem       thread's own semaphore
+ * @param other_sem    other thread's semaphore
+ */
+void helloLoop(const char *my_name, struct k_sem *my_sem,
+	struct k_sem *other_sem, int incrementer)
+{
+	const char *tname;
+	uint8_t cpu = 0;
+	struct k_thread *current_thread;
+	int task_local_counter = 0;
+
+	while (1) {
+		/* take my semaphore */
+		k_sem_take(my_sem, K_FOREVER);
+
+		k_spinlock_key_t key1 = k_spin_lock(&lock1);
+		{
+			current_thread = k_current_get();
+			tname = k_thread_name_get(current_thread);
+#if defined(CONFIG_ARC) || defined(CONFIG_ARM) || defined(CONFIG_X86_64) \
+	|| defined(CONFIG_RISCV) || defined(CONFIG_XTENSA)
+			cpu = arch_curr_cpu()->id;
+#endif
+			printk("%s: level 1 spinlock key = %d\n", tname, key1.key);
+			k_spinlock_key_t key2 = k_spin_lock(&lock2);
+			{
+				k_yield();
```

Comment:
What is the reason for doing switch while holding lock?

This really looks like incorrect thing by multiple reasons. Inside of spinlock we disable interrupts so after the switch we appear in another thread with interrupt disabled (and no proper way to enable them back).

In case of SMP it's even worse - we may need to do ICI (inter-core-interrupt) from scheduler code - but we may have interrupts disabled on a core.

Just to pove that it's not allowed - if we enable asserts (CONFIG_ASSERT=y) we'll get get assertion in kernel code about it (and system will halt):

```
*** Booting Zephyr OS build v3.4.0-508-g5c1682115388 ***
thread_a: level 1 spinlock key = 49
ASSERTION FAIL [arch_irq_unlocked(key) || _kernel.cpus[0].current->base.thread_state & (((1UL << (0))) | ((1UL << (3))))] @ WEST_TOPDIR/zephyr/kernel/include/kswap.h:99
        Context switching while holding lock!
```

What we are trying to protect with these spinlocks here? Can we drop them?

Reply:
> CONFIG_ASSERT=y
I've learnt something new. Thank you!
Short answer; original idea was the to disable interrupts and yield just for the kicks.
Long answer; to prove the behaviour under "weird" code. Because context switch implementations (z_swap etc.) has an interrupt status restore mechanism that I wanted to test.
Yes we can drop them. They're not crucial to the no-timer test, but do you see any issues with at least 1 level of `irq_lock`s ?

Reply:
Yes there is an issue. it's the same. I've dropped any attempt to switch context under irq locks.